text/0000-rand-crate-redesign.md

```diff
@@ -0,0 +1,1243 @@
+required for the special `Weighted` distribution, which does not support `Copy`.
+Does this add overhead? Note that currently `rand` is implemented using
+`sample`, which in some ways is the worst of both worlds. Should distributions
+be required to support `Copy` or, at least, should `sample` take `self` by
+value?
+
+### Distributions trait
+
+The [`Distribution`] trait replaces `rand`'s current [`IndependentSample`]
+trait. It is quite simple:
+
+```rust
+/// Types (distributions) that can be used to create a random instance of `T`.
+pub trait Distribution<T> {
+    /// Generate a random value of `T`, using `rng` as the
+    /// source of randomness.
+    fn sample<R: Rng+?Sized>(&self, rng: &mut R) -> T;
+}
+```
+
+This could be extended with other functions such as
+`fn map<F: Fn(T) -> T>(&self, f: F) -> MappedDistribution<T, F>`, but I do not
+see a good rationale.
+
+Any implementation, such as [`Default`], supports usage via `sample`:
+`Default.sample(&mut rng)`. (Note that `struct Default;` is a unit type (like `()`); Rust
+allows objects to be created without any extra syntax: `let x = Default;`.)
+
+#### Simple distributions
+
+Several zero-size structs implementing [`Distribution`] specify simple distributions:
+
+*   [`Uniform`] specifies uniform distribution over the entire range available, and
+    is implemented for all integer types and `bool`
+*   [`Uniform01`] specifies uniform distribution over the half-open range `[0, 1)`,
+    and is implemented for `f32` and `f64`
+*   [`Closed01`] is like [`Uniform01`] but for `[0, 1]` (thus including 1.0)
+*   [`Open01`] is like [`Uniform01`] but for `(0, 1)` (thus excluding 0.0)
+*   [`Default`] uses [`Uniform`] or [`Uniform01`] depending on type (and can be
+    extended for other types)
+*  [`AsciiWordChar`] samples uniformly from the ASCII characters 0-9, A-Z and a-z
+
+[`Default`] has roughly the same capabilities as the the current `rand` crate's
+[`Rand`](https://docs.rs/rand/0.3.15/rand/trait.Rand.html); currently it doesn't
+support arrays, tuples, `Option`, etc., but support for thes could conceivably
+be added, and probably also an equivalent to `derive_rand`.
+
+It should be noted that there is no agreement on using the name `Default`. In
+particular, there is a naming conflict with `std::default::Default`, which can
+lead to surprising compiler messages if the user forgets to
+`use rand::Default;`.
+Potentially `Default` could be renamed to `Rand`; my personal feeling is that
+the name `Default` works well.
+
+Similarly, `Uniform` and `Uniform01` are open to
+adjustment. All three (including `Default`) could be replaced with a single
+`Uniform`; but using three names does solve
+two semantic issues: (1) the range of sampled values differs by type, especially
+between integer and floating-point types, and (2) some possible
+type-dependent implementations (such as for `Option`) cannot practically have
+a uniform distribution.
```

Comment:
I like the split between `Uniform`, `Uniform01` and `Default`. Not all distributions make sense for all types. Having a `Default` distribution seems like a nice solution. A thesaurus does not really help to find a different name, but maybe `Basic`? Thinking about it for a few minutes, I start to like it.